sdk/python/lib/pulumi/provider/server.py

```diff
@@ -0,0 +1,189 @@
+        urn = await pulumi.Output.from_input(result.urn).future()
+
+        # Note: property_deps is populated by rpc.serialize_properties.
+        property_deps: Dict[str, List[pulumi.resource.Resource]] = {}
+        state = await rpc.serialize_properties(
+            inputs={k: v for k, v in result.state.items() if k not in ['id', 'urn']},
+            property_deps=property_deps)
+
+        deps: Dict[str, proto.ConstructResponse.PropertyDependencies] = {}
+        for k, resources in property_deps.items():
+            urns = await asyncio.gather(*(r.urn.future() for r in resources))
+            deps[k] = proto.ConstructResponse.PropertyDependencies(urns=urns)
+
+        return proto.ConstructResponse(urn=urn,
+                                       state=state,
+                                       stateDependencies=deps)
+
+    async def Configure(self, request, context) -> proto.ConfigureResponse:  # pylint: disable=invalid-overridden-method
+        return proto.ConfigureResponse(acceptSecrets=True, acceptResources=True)
+
+    async def CheckConfig(self, request, context) -> proto.CheckResponse:  # pylint: disable=invalid-overridden-method
+        # NOTE: inputs=None here seems to work but we may be required
+        # to remember and mirror back the inputs passed in Configure
+        #
+        # google.protobuf.Struct inputs = 1; // the provider inputs for this resource.
+        return proto.CheckResponse(inputs=None, failures=[])
+
+    async def GetPluginInfo(self, request, context) -> proto.PluginInfo:  # pylint: disable=invalid-overridden-method
+        return proto.PluginInfo(version=self.provider.version)
+
+    def __init__(self, provider: Provider, args: List[str], engine_address: str) -> None:
+        super().__init__()
+        self.provider = provider
+        self.args = args
+        self.engine_address = engine_address
+
+
+
+def main(provider: Provider, args: List[str]) -> None:  # args not in use?
+    """For use as the `main` in programs that wrap a custom Provider
+    implementation into a Pulumi-compatible gRPC server.
+
+    :param provider: an instance of a Provider subclass
+
+    :args: command line arguiments such as os.argv[1:]
+
+    """
+
+    argp = argparse.ArgumentParser(description='Pulumi provider plugin (gRPC server)')
+    argp.add_argument('engine', help='Pulumi engine address')
+    engine_address: str = argp.parse_args().engine
+
+    async def serve() -> None:
+        server = grpc.aio.server(options=_GRPC_CHANNEL_OPTIONS)
+        servicer = ProviderServicer(provider, args, engine_address=engine_address)
+        provider_pb2_grpc.add_ResourceProviderServicer_to_server(servicer, server)
+        port = server.add_insecure_port(address='0.0.0.0:0')
+        await server.start()
+        sys.stdout.buffer.write(f'{port}\n'.encode())
+        sys.stdout.buffer.flush()
+        await server.wait_for_termination()
+
+    try:
+        asyncio.run(serve())
```

Comment:
We can't use `asyncio.run` as it was only introduced in python 3.7 and we support python 3.6. You'll have to use `asyncio.get_event_loop().run_until_complete(serve())` or something along those lines.

Reply:
Hmm. I can def make that fix but I'm not confident the entire thing runs on 3.6. Do we have a way to run automated tests against 3.6?

Reply:
We definitely have users that are on 3.6 and they certainly [bring it up](https://github.com/pulumi/pulumi/issues/6108) when there's a break. We also go to [great pains](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/runtime/sync_await.py#L20-L38) to make sure we remain backward compatible with 3.6. 

We don't have automated testing with 3.6, we've tried implementing it but have run into issues with the test runners themselves, and not the actual SDK. [There's an open issue](https://github.com/pulumi/pulumi/issues/6110).

Reply:
Thanks for the context. I'll implement your suggestion and watch out for any complaints.

Reply:
At the very least, if we are choosing to break with our existing [supported languages](https://www.pulumi.com/docs/intro/languages/#languages) the decision should be explicit and agreed upon by the team.